admin-ui/src/main/java/io/crate/rest/action/admin/AdminUIFrontpageAction.java

```diff
@@ -84,4 +99,106 @@ private boolean isBrowser(String headerValue) {
 
         return matcher.matches();
     }
+
+    private class AdminUIRequestFilter extends RestFilter {
+
+        @Override
+        public void process(RestRequest request, RestChannel channel, RestFilterChain filterChain) throws IOException {
+            if(request.rawPath().startsWith(ADMIN_ENPOINT)) {
+                serveSite(request, channel);
+            } else {
+                filterChain.continueProcessing(request, channel);
+            }
+        }
+    }
+
+    private void serveSite(RestRequest request, RestChannel channel) throws IOException {
+        if (request.method() != RestRequest.Method.GET) {
+            channel.sendResponse(new BytesRestResponse(FORBIDDEN));
+            return;
+        }
+
+        String sitePath = request.rawPath().substring(ADMIN_ENPOINT.length());
+
+        // we default to index.html, or what the plugin provides (as a unix-style path)
+        // this is a relative path under _site configured by the plugin.
+        if (sitePath.length() == 0) {
+            sitePath = "index.html";
+        } else {
+            // remove extraneous leading slashes, its not an absolute path.
+            while (sitePath.length() > 0 && sitePath.charAt(0) == '/') {
+                sitePath = sitePath.substring(1);
+            }
+        }
+        final Path siteFile = environment.pluginsFile().resolve("crate-admin").resolve("_site");
+
+        final String separator = siteFile.getFileSystem().getSeparator();
+        // Convert file separators.
+        sitePath = sitePath.replace("/", separator);
+
+        Path file = siteFile.resolve(sitePath);
+
+        // return not found instead of forbidden to prevent malicious requests to find out if files exist or dont exist
+        if (!Files.exists(file) || FileSystemUtils.isHidden(file) || !file.toAbsolutePath().normalize().startsWith(siteFile.toAbsolutePath().normalize())) {
+            channel.sendResponse(new BytesRestResponse(NOT_FOUND));
+            return;
+        }
+
+        BasicFileAttributes attributes = Files.readAttributes(file, BasicFileAttributes.class);
+        if (!attributes.isRegularFile() && !attributes.isDirectory()) {
+            // If it's not a dir, we send a 403
```

Comment:
why should we allow directory access? afaik this is not needed and any directory access should result in a 403